rrdd/rrdd_server.ml

```diff
@@ -275,18 +275,66 @@ let forget_host_ds _ ~(ds_name : string) : unit =
 	)
 
 let query_possible_dss rrdi =
-	let enabled_dss = Rrd.ds_names rrdi.rrd in
-	let open Ds in
-	let open Data_source in
-	List.map (fun ds -> {
-		name = ds.ds_name;
-		description = ds.ds_description;
-		enabled = List.mem ds.ds_name enabled_dss;
-		standard = ds.ds_default;
-		min = ds.ds_min;
-		max = ds.ds_max;
-		units = ds.ds_units;
-	}) rrdi.dss
+  (* We have data sources coming from different places, so we want the union of these;
+     this happens to be the union of the rrdi.rrd (live) and rrdi.rrd.rdd_dss (archival).
+     This is straighforward, but those two sets of ds have different types,
+     hence we need to coerce them into a common type (this is why we have
+     'description not available', and 'units unknown' etc.).
+     Deciding whether a ds is enabled is also not obvious. If we have a 'live'
+     ds, then it is enabled if it exists in the set rrdi.rrd. If we have an 'archival'
+     ds, then it is enabled if it is also an enabled 'live' ds, otherwise it is
+     disabled.
+   *)
+  let module SSet = Set.Make(String) in
+  let module SMap = Map.Make(String) in
+  let add_array xs f m =
+    Array.fold_left (fun m x -> match f m x with
+      | None        -> m
+      | Some (k, v) -> SMap.add k v m) m xs
+  in
+  let open Ds in
+  let open Data_source in
+
+  let live_sources = rrdi.dss in
+  let archival_sources = rrdi.rrd.rrd_dss in
+
+  let name_to_live_dss =
+    let enabled_names = Rrd.ds_names rrdi.rrd |> SSet.of_list in
+    let is_live_ds_enabled ds = SSet.mem ds.ds_name enabled_names in
+    live_sources |>
+    List.fold_left (fun map ds ->
+      let k = ds.ds_name in
+      let v = { name = ds.ds_name
+              ; description = ds.ds_description
+              ; enabled = is_live_ds_enabled ds
+              ; standard = ds.ds_default
+              ; min= ds.ds_min
+              ; max = ds.ds_max
+              ; units = ds.ds_units
+              }
+      in
+      SMap.add k v map) SMap.empty
+  in
+  let archival_source_to_kv m ds =
+    if SMap.mem ds.Rrd.ds_name m then
+      None
+    else
+      let k = ds.ds_name in
+      let v = { name = ds.ds_name
+              ; description = "description not available"
+              ; enabled = false
+              ; standard = false
+              ; units = "unknown"
+              ; min = ds.ds_min
+              ; max = ds.ds_max
+              }
+      in
+      Some (k,v)
+  in
+  SMap.add_array archival_sources archival_source_to_kv name_to_live_dss |>
```

Comment:
I find the usage of pattern matching makes the code tougher to follow that what it could be, compare it with:
```
let add_if_source_missing map ds = 
  if not Smap.mem ds.Rrd.ds_name map then
    let k = ds.ds_name in
    let v = { name = ds.ds_name
             ; ... }
    in
    Smap.add k v map
  else
    map
in
let add_missing_sources name_to_dss sources = 
  Array.fold_left add_if_source_missing name_to_dss sources
in
add_missing_sources name_to_live_dss archival_sources |> ...
```